### ISSUE TITLE
How to control deletion of rows in table?

### ISSUE THREAD

**Issue by bonjefir** [gregnb/mui-datatables#193]
I need to cancel deletion of some of row(s) which user deleted but its operation failed in back-end side for some reason.

```
const options = {
	filter: true,
	filterType: 'dropdown',
	responsive: 'scroll',
	onRowsDelete: (rowsDeleted) => {
		for (var key in rowsDeleted.data) {
			this.removeItem(this.state.item_id[rowsDeleted.data[key].dataIndex])
				.then(res => {
					if (res != true) {
						// ???
					}
			})
		}
		console.log(rowsDeleted, "were deleted!");
	}
};
```
Here `removeItem` returns a promise which shows the status of deletion. I am wondering deletion in UI applied after calling `onRowsDelete` or before it and how can recover deleted rows OR preventing MUI to delete items in its `rowsDeleted` which their `res` is `false`.

**Comment by gregnb** [comment]
Hey @bonjefir thanks for filing the issue. Let me think about this some more. 

One change that will soon happen is the rowDelete action will be added to the onServerRequest callback so you can look for that event there and react to it. If you do receive the event there, maybe if there is a total failure you can just return the data for that page back to what it was from your API?

**Comment by majose1500** [comment]
Sorry but the onRowsDelete options not work to me. I'm trying to do a basic example like show message when a row was delete but do not see nothing. Exist any functional example of onRowsDelete options?

`const options = {
	filterType: 'dropdown',
	responsive: 'stacked',
	onRowsDelete: (rowsDeleted) => {
		console.log(rowsDeleted, "were deleted!");
	}
};`

**Comment by toolaugh** [comment]
Normaly, If we want to delete a row of table, We have to pass a ID of the data Row and send it to the server. But, The data that we use to build the table does not have the ID of the row data. So we have to map the selected row with the data from server that we want to delete.

**Comment by d-gubert** [comment]
> Normaly, If we want to delete a row of table, We have to pass a ID of the data Row and send it to the server. But, The data that we use to build the table does not have the ID of the row data. So we have to map the selected row with the data from server that we want to delete.

To solve this you could add the ID as the first column, and not actually show the column, like this:
```javascript
columns = [
    {
        name: "ID",
        options: {
            display: "excluded",
            filter: false,
            sort: false,
            download: false,
        },
    },
    "Pacote",
    "Validade",
    "Data de Criação",
    "Ativo",
];
```
Then normally add the record's id to your `data`, so that you're able to retrieve it on the `onRowsDeleted` event

**Comment by matteo1976** [comment]
sorry but if i did onRowsDelete: (rowsDeleted) => { console.log(rowsDeleted, "were deleted!"); } };
I have in return :
Object {lookup: Object, data: Array[1]}
where data is array 
      data: Array[1]
only with this data
       index: 1
       dataIndex: 2

My question is.... now how can I use this information to delete my data in remote? I need to have back from onRowsDelete an object that have the data of all my columns

**Comment by gabrielliwerant** [comment]
I've got a PR up that might help. The idea is that returning `false` from `onRowsDelete` will prevent a delete action from taking place. There will be an example in the code to demonstrate this as well.

**Comment by gabrielliwerant** [comment]
The ability to skip deletion of a row has been available since `2.1.0` by returning `false` from `onRowsDelete`.

**Comment by leiyang** [comment]
> 
> 
> sorry but if i did onRowsDelete: (rowsDeleted) => { console.log(rowsDeleted, "were deleted!"); } };
> I have in return :
> Object {lookup: Object, data: Array[1]}
> where data is array
> data: Array[1]
> only with this data
> index: 1
> dataIndex: 2
> 
> My question is.... now how can I use this information to delete my data in remote? I need to have back from onRowsDelete an object that have the data of all my columns

Hi  gabrielliwerant,  did you get the answer ? 

Lei
